src/dfcx_scrapi/tools/evaluations.py

```diff
@@ -268,6 +269,154 @@ def run_query_and_eval(self, df: pd.DataFrame) -> pd.DataFrame:
+                                else:
+                                    input_params[field.key] = field.value.string_value if hasattr(field.value, 'string_value') else field.value.number_value
+                        
+                        output_param = action.tool_use.output_action_parameters
+                        if hasattr(output_param, 'fields'):
+                            for field in output_param.fields:
+                                output_params[field.key] = field.value.string_value if hasattr(field.value, 'string_value') else field.value.number_value
+                        
+                        interaction_result["tool_calls"].append({
+                            "tool_name": matched_tool_name,
+                            "action": action_name,
+                            "input_parameters": input_params,
+                            "output_parameters": output_params
+                        })
+                    
+                    elif "agent_utterance" in action:
+                        response_text = action.agent_utterance.text
+                        interaction_result["responses"].append(response_text)
+                    
+                    elif "playbook_invocation" in action:
+                        playbook_url = action.playbook_invocation.playbook
+                        if isinstance(playbooks_map, dict):
+                            matched_playbooks_name = playbooks_map.get(playbook_url)
+                        interaction_result["playbook_invocation"] = matched_playbooks_name
+            
+            results.append(interaction_result)
+        
+        return results
+    
+    def create_golden_dataset_from_conv_ids(self, conversation_ids):
+        ch = ConversationHistory()
+        
+        INPUT_SCHEMA_REQUIRED_COLUMNS = [
+            'eval_id', 'action_id', 'action_type', 'action_input', 
+            'action_input_parameters', 'tool_action', 'notes'
+        ]
+        df_rows = []
+        
+        for idx, conv_id in enumerate(conversation_ids, start=1):
+            eval_id = f"{idx:03d}"
+            
+            convo = ch.get_conversation(conv_id)
+            parsed_conv = self.parse_interactions_from_conversation_history(convo)
+            
+            action_counter = 1
+            
+            playbook_count = sum(1 for interaction in parsed_conv if 'playbook_invocation' in interaction)
+            
+            for interaction in parsed_conv:
+                df_rows.append({
+                    'eval_id': eval_id,
+                    'action_id': action_counter,
+                    'action_type': 'User Utterance',
+                    'action_input': interaction['query'],
+                    'action_input_parameters': '',
+                    'tool_action': '',
+                    'notes': ''
+                })
+                action_counter += 1
```

Comment:
This code is duplicated several times in the same method.
You could break this out into a single method like `append_to_rows()` and then pass in the needed variables to make the append happen